File: src/lexer.cpp
#include "lexer.hpp"
#include "token.hpp"

#include <cassert>
#include <cstdlib>
#include <utility>

Lexer::Lexer(string_view src) { this->src = src; }

char Lexer::current() { return this->src.at(this->location); }

void Lexer::advance() { this->location++; }

bool Lexer::advance_if(char c) {
    if (this->match(c)) {
        this->advance();
        return true;
    }

    return false;
}

bool Lexer::match(char c) { return this->current() == c; }

bool Lexer::eof() { return this->location >= this->src.size(); }

void Lexer::append_token(TokenType type) {
    this->out.push_back(Token{
        .type = type,
        .location = this->start,
        .line = this->line,
        .line_offset = this->line_offset,
        .text = this->src.substr(this->start, this->location - this->start),
    });
}

void Lexer::append_ident() {
    auto text = this->src.substr(this->start, this->location - this->start);

    TokenType type = string_to_tokentype(text);

    this->out.push_back(Token{
        .type = type,
        .location = this->start,
        .line = this->line,
        .line_offset = this->line_offset,
        .text = text,
    });
}

void Lexer::append_single(TokenType type) {
    this->advance();
    this->append_token(type);
}

void Lexer::append_single_or_next(char cond, TokenType t, TokenType f) {
    this->advance();
    this->append_token(this->advance_if(cond) ? t : f);
}

static bool is_letter(char c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
}

static bool is_digit(char c) { return (c >= '0' && c <= '9'); }

static bool is_ident_start_char(char c) { return is_letter(c) || c == '_'; }
static bool is_ident_middle_char(char c) {
    return is_ident_start_char(c) || is_digit(c);
}

std::vector<Token> Lexer::run() {
    while (!this->eof()) {
        char c = this->current();
        this->start = this->location;

        switch (c) {
        case '(': {
            this->append_single(TokenType::LParen);
        } break;
        case ')': {
            this->append_single(TokenType::RParen);
        } break;
        case '{': {
            this->append_single(TokenType::LBrace);
        } break;
        case '}': {
            this->append_single(TokenType::RBrace);
        } break;
        case '[': {
            this->append_single(TokenType::LBracket);
        } break;
        case ']': {
            this->append_single(TokenType::RBracket);
        } break;

        case ';': {
            this->append_single(TokenType::SemiColon);
        } break;
        case ':': {
            this->append_single(TokenType::Colon);
        } break;

        case '.': {
            this->append_single(TokenType::Dot);
        } break;
        case ',': {
            this->append_single(TokenType::Comma);
        } break;

        case '+': {
            this->append_single(TokenType::Plus);
        } break;
        case '-': {
            this->append_single(TokenType::Minus);
        } break;
        case '*': {
            this->append_single(TokenType::Multiply);
        } break;
        case '/': {
            this->append_single(TokenType::Divide);
        } break;

        case '=': {
            this->append_single_or_next('=', TokenType::EqualEq,
                                        TokenType::Equal);
        } break;
        case '!': {
            this->append_single_or_next('=', TokenType::NotEq, TokenType::Not);
        } break;

        case '"': {
            this->advance();
            while (!this->match('"')) {
                this->advance();
            }

            this->advance();
            this->append_token(TokenType::String);
        } break;

        case '\n':
        case '\r': {
            this->line++;
            this->line_offset = 0;
        }
        case '\t':
        case ' ': {
            this->advance();
        } break;

        default: {
            if (is_ident_start_char(c)) {
                this->advance();

                while (is_ident_middle_char(this->current())) {
                    this->advance();
                }

                this->append_ident();
                continue;
            }

            if (is_digit(c)) {
                this->advance();

                bool is_float = false;
                while (is_digit(this->current()) || this->current() == '.') {
                    if (this->match('.'))
                        is_float = true;

                    this->advance();
                }

                this->append_token(is_float ? TokenType::Float
                                            : TokenType::Int);
                continue;
            }

            fmt::println("Current char: '{}'", this->current());
            abort();
        } break;
        }
    }

    return std::move(this->out);
}

File: src/lexer.rs
use crate::tokens::{self, Token, TokenType};

pub struct Lexer<'a> {
    start: usize,
    position: usize,
    line: usize, // TODO: Add line offset

    src: &'a [u8],
    out: Vec<Token<'a>>,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str) -> Self {
        Self {
            start: 0,
            position: 0,
            line: 1,

            src: src.as_bytes(),

            out: vec![],
        }
    }

    fn current(&self) -> u8 {
        self.src[self.position]
    }

    fn append(&mut self, ty: TokenType) {
        let text = unsafe { std::str::from_utf8_unchecked(&self.src[self.start..self.position]) };
        let ty = if ty == TokenType::Ident {
            match tokens::KEYWORDS.get(text).cloned() {
                Some(ty) => ty,
                None => TokenType::Ident,
            }
        } else {
            ty
        };

        let token = Token {
            text,
            line: self.line,
            location: self.start,
            ty,
        };

        self.out.push(token);
    }

    fn append_single(&mut self, ty: TokenType) {
        self.advance();
        self.append(ty);
    }

    fn append_single_or_next(&mut self, cond: u8, t: TokenType, f: TokenType) {
        self.advance();
        let tok = if self.advance_if(cond) { t } else { f };
        self.append(tok);
    }

    fn matches(&self, c: u8) -> bool {
        self.current() == c
    }

    fn advance(&mut self) {
        self.position += 1;
    }

    fn advance_if(&mut self, cond: u8) -> bool {
        if self.matches(cond) {
            self.advance();
            true
        } else {
            false
        }
    }

    pub fn run(mut self) -> Vec<Token<'a>> {
        while self.position < self.src.len() {
            self.start = self.position;

            match self.current() {
                b'(' => self.append_single(TokenType::LParen),
                b')' => self.append_single(TokenType::RParen),
                b'{' => self.append_single(TokenType::LBrace),
                b'}' => self.append_single(TokenType::RBrace),
                b'[' => self.append_single(TokenType::LBracket),
                b']' => self.append_single(TokenType::RBracket),

                b'.' => self.append_single(TokenType::Dot),
                b',' => self.append_single(TokenType::Comma),

                b':' => self.append_single(TokenType::Colon),
                b';' => self.append_single(TokenType::Semicolon),

                b'+' => self.append_single_or_next(b'=', TokenType::PlusEq, TokenType::Plus),
                b'-' => self.append_single_or_next(b'=', TokenType::MinusEq, TokenType::Minus),
                b'*' => self.append_single_or_next(b'=', TokenType::MulEq, TokenType::Asterix),
                b'/' => {
                    self.advance();

                    match self.current() {
                        b'/' => {
                            while !self.matches(b'\n') {
                                self.advance();
                            }
                        }
                        b'*' => todo!("Add multiline comments"),
                        b'=' => self.append_single(TokenType::DivEq),
                        _ => self.append(TokenType::Slash),
                    }
                }

                b'=' => self.append_single_or_next(b'=', TokenType::EqualEqual, TokenType::Equal),
                b'!' => self.append_single_or_next(b'=', TokenType::NotEqual, TokenType::Not),

                b'"' => {
                    self.advance();

                    while self.current() != b'"' {
                        self.advance();
                    }

                    self.start += 1;
                    self.append(TokenType::String);

                    self.advance();
                }

                b' ' | b'\t' => self.advance(),
                b'\n' | b'\r' => {
                    self.advance();
                    self.line += 1;
                }

                c if c.is_ascii_digit() => {
                    self.advance();

                    let mut ty = TokenType::Int;

                    loop {
                        match self.current() {
                            b'.' => ty = TokenType::Float,
                            c if c.is_ascii_digit() => {}
                            _ => break,
                        }

                        self.advance();
                    }

                    self.append(ty);
                }
                c if c.is_ascii_alphabetic() || c == b'_' => {
                    self.advance();

                    loop {
                        match self.current() {
                            b'_' => {}
                            c if c.is_ascii_alphanumeric() => {}
                            _ => break,
                        }

                        self.advance();
                    }

                    self.append(TokenType::Ident);
                }

                c => panic!(
                    "Not implemented for character '{}'",
                    std::str::from_utf8(std::slice::from_ref(&c)).unwrap()
                ),
            }
        }

        self.out
    }
}
